slack/webhook/client.py

```diff
@@ -0,0 +1,125 @@
+        """
+        self.url = url
+        self.default_headers = default_headers
+
+    def send(
+        self, body: Dict[str, any], additional_headers: Dict[str, str] = {},
+    ) -> WebhookResponse:
+        """Performs a Slack API request and returns the result.
+        :param url: a complete URL (e.g., https://hooks.slack.com/XXX)
+        :param json_body: json data structure (it's still a dict at this point),
+            if you give this argument, body_params and files will be skipped
+        :param body_params: form params
+        :param additional_headers: request headers to append
+        :return: API response
+        """
+
+        body = convert_bool_to_0_or_1(body)
+        self._parse_blocks(body)
+        if self.logger.level <= logging.DEBUG:
+            self.logger.debug(
+                f"Slack API Request - url: {self.url}, "
+                f"body: {body}, "
+                f"additional_headers: {additional_headers}"
+            )
+
+        request_headers = self._build_request_headers(
+            has_json=json is not None, additional_headers=additional_headers,
+        )
+        args = {
+            "headers": request_headers,
+            "body": body,
+        }
+        return self._perform_http_request(url=self.url, args=args)
+
+    def _perform_http_request(
+        self, *, url: str, args: Dict[str, Dict[str, any]]
+    ) -> WebhookResponse:
+        """Performs an HTTP request and parses the response.
+        :param url: a complete URL (e.g., https://www.slack.com/api/chat.postMessage)
+        :param args: args has "headers", "data", "params", and "json"
+            "headers": Dict[str, str]
+            "params": Dict[str, str],
+            "json": Dict[str, any],
+        :return: a tuple (HTTP response and its body)
+        """
+        headers = args["headers"]
+        body = json.dumps(args["body"]).encode("utf-8")
+        headers["Content-Type"] = "application/json;charset=utf-8"
+
+        try:
+            if url.lower().startswith("http"):
+                req = Request(method="POST", url=url, data=body, headers=headers)
+            else:
+                raise SlackRequestError(f"Invalid URL detected: {url}")
+            resp: HTTPResponse = urlopen(req)
+            charset = resp.headers.get_content_charset() or "utf-8"
+            return WebhookResponse(
+                url=self.url,
+                status_code=resp.status,
+                body=resp.read().decode(charset),
+                headers=resp.headers,
+            )
+        except HTTPError as e:
+            charset = e.headers.get_content_charset() or "utf-8"
+            resp = WebhookResponse(
+                url=self.url,
+                status_code=e.code,
+                body=e.read().decode(charset),
+                headers=e.headers,
+            )
+            if e.code == 429:
+                resp.headers["Retry-After"] = resp.headers["retry-after"]
```

Comment:
Is this due to rate limiting?

Reply:
Yes, it is. 

The purpose of this line is to have backward-compatibility with `WebClient` in older versions (2.5.0 or older). Although this class doesn't rely on aiohttp at all, accessing the value by this key is mentioned in our `WebClient` document. Developers may expect `WebhookClient` the same behavior.